src/main/java/org/broadinstitute/hellbender/tools/reference/ReferenceSequenceTable.java

```diff
@@ -0,0 +1,437 @@
+                }
+                tableByMD5.get(md5).add(newEntry);
+
+                if(!tableBySequenceName.containsKey(name)){
+                    Set<TableRow> rowSet = new HashSet<>();
+                    tableBySequenceName.put(name, rowSet);
+                }
+                tableBySequenceName.get(name).add(tableByMD5.get(md5));
+            }
+        }
+        tableBuilt = true;
+    }
+
+    public Set<String> getAllSequenceNames(){
+        checkTableBuildStatus();
+
+        return tableBySequenceName.keySet();
+    }
+
+    /**
+     * Given an MD5, returns its corresponding row
+     *
+     * @param md5 The MD5 as a String
+     * @return the corresponding TableRow from the tableByMD5
+     */
+    public TableRow queryByMD5(String md5){
+        checkTableBuildStatus();
+
+        return tableByMD5.get(md5);
+    }
+
+    /**
+     * Given a sequence name, returns the set of its corresponding rows or an empty set if no such rows found.
+     *
+     * @param sequenceName The sequence name as a String
+     * @return the set of TableRows that contain the sequence name
+     */
+    public Set<TableRow> queryBySequenceName(String sequenceName){
+        checkTableBuildStatus();
+
+        Set<TableRow> rows = tableBySequenceName.get(sequenceName);
+        return rows == null ? Collections.emptySet() : rows;
+    }
+
+    private String getMD5ForRecord(SAMSequenceRecord record, GATKPath referencePath) {
+        final String md5FromDict = record.getMd5();
+        final SimpleInterval referenceInterval = new SimpleInterval(record.getSequenceName(), 1, record.getSequenceLength());
+        String md5 = null;
+
+        switch (md5Mode) {
+            case USE_DICT:
+                if (md5FromDict == null || md5FromDict.isEmpty()) {
+                    throw new UserException.BadInput("Running in USE_DICT mode, but MD5 missing for sequence " + record.getSequenceName() + ". Run --md5-calculation-mode with a different mode to recalculate MD5.");
+                }
+                md5 = md5FromDict;
+                break;
+            case RECALCULATE_IF_MISSING:
+                if (md5FromDict == null || md5FromDict.isEmpty()) {
+                    logger.warn(String.format("%s in sequence dictionary for %s missing MD5. MD5 recalculated.", record.getSequenceName(), getReferenceColumnName(referencePath)));
+                    md5 = ReferenceUtils.calculateMD5(referencePath, referenceInterval);
+                }
+                else {
+                    md5 = md5FromDict;
+                }
+                break;
+            case ALWAYS_RECALCULATE:
+                md5 = ReferenceUtils.calculateMD5(referencePath, referenceInterval);
+                if(md5FromDict != null && !md5FromDict.equals(md5)){
+                    logger.warn(String.format("MD5 Mismatch for sequence %s. Found '%s', but calculated '%s'. Sequence dictionary may be invalid.", record.getSequenceName(), md5FromDict, md5));
```

Comment:
This warning should mention the name of the reference involved. It should also say "Found %s in sequence dictionary" for clarity